src/Radicle/Internal/Pretty.hs

```diff
@@ -17,43 +20,49 @@ import qualified Radicle.Internal.Annotation as Ann
 import           Radicle.Internal.Core
 import           Radicle.Internal.Type
 
+class PrettyV a where
+  prettyV :: a -> Doc Type
+
 -- | We can't just pretty print the pointer id since that would break
 -- referential transparency, so instead we just label refs as '<ref>'
-instance Pretty Reference where
-    pretty _ = angles "ref"
+instance PrettyV Reference where
+    prettyV _ = annotate TRef $ angles "ref"
 
 instance Pretty Ident where
     pretty (Ident i) = pretty i
 
-instance Pretty Type where
-  pretty = pretty . typeToValue
-
-instance (Copointed t, Ann.Annotation t) => Pretty (Ann.Annotated t ValueF) where
-    pretty v = case v of
-        Atom i -> pretty i
-        Keyword i -> ":" <> pretty i
-        Ref i -> pretty i
-        String t -> "\"" <> pretty (escapeStr t) <> "\""
-        Number (a :% b) -> pretty a <> if b == 1 then "" else "/" <> pretty b
-        Boolean True -> "#t"
-        Boolean False -> "#f"
-        List vs -> case vs of
+instance PrettyV Type where
+  prettyV = prettyV . typeToValue
+
+instance forall t. (Copointed t, Ann.Annotation t) => PrettyV (Ann.Annotated t ValueF) where
+    prettyV v = case v of
+        Atom i -> annotate TAtom $ pretty i
+        Keyword i -> annotate TKeyword $ ":" <> pretty i
+        Ref i -> prettyV i
+        String t -> annotate TString $ "\"" <> pretty (escapeStr t) <> "\""
+        Number (a :% b) -> annotate TNumber $ pretty a <> if b == 1 then "" else "/" <> pretty b
+        Boolean True -> annotate TBoolean $ "#t"
+        Boolean False -> annotate TBoolean $ "#f"
+        List vs -> annotate TList $ case vs of
           []   -> "()"
-          [v'] -> parens $ pretty v'
-          _    -> parens $ hang 1 (sep $ pretty <$> vs)
-        Vec vs -> case vs of
+          [v'] -> parens $ prettyV v'
+          _    -> parens $ hang 1 (sep $ prettyV <$> vs)
+        Vec vs -> annotate TVec $ case vs of
           Empty        -> "[]"
-          v' :<| Empty -> brackets $ pretty v'
-          _            -> brackets . sep $ pretty <$> toList vs
-        PrimFn i -> pretty i
-        Dict mp -> braces . align $
-            sep [ pretty k <+> pretty val
+          v' :<| Empty -> brackets $ prettyV v'
+          _            -> brackets . sep $ prettyV <$> toList vs
+        PrimFn i -> annotate TFunction $ pretty i
+        Dict mp -> annotate TDict $ braces . align $
+            sep [ prettyV k <+> prettyV val
                 | (k, val) <- Map.toList mp ]
-        Lambda ids vals _ -> parens $
-            "fn" <+> align (sep
-                        [ brackets . sep $ pretty <$> ids
-                        , sep $ pretty <$> toList vals
-                        ])
+        Lambda ids vals _ ->
+          let args :: Ann.Annotated t ValueF = Vec $ Seq.fromList (Atom <$> ids) in
+          annotate TFunction $ parens $
```

Comment:
This section is a little confusingly formatted. Can the `in` be put on the next line, and arguments to applications indented?